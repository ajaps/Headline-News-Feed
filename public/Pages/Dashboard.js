import React from 'react';

import Footer from '../component/Footer';
import Header from '../component/Header';
import Navbar from '../component/HeadlineNavbar';

import ArticleComponent from '../component/ArticleComponent';
import SourcesComponent from '../component/SourcesComponent';

import ArticleStore from '../stores/Article';
import SourcesStore from '../stores/Sources';

import * as myActions from '../actions/HeadlineActions';


export default class Dashboard extends React.Component {
  constructor() {
    super();
    this.getSources();
    this.setSources = this.setSources.bind(this);
    this.setArticles = this.setArticles.bind(this);
    this.state = { title: 'Ajaps Franklin that BOSS!!',
      article: ArticleStore.getAll(),
      sources: SourcesStore.getAll(),
      searchTerm: '',
    };
  }

  componentWillMount() {
    SourcesStore.on('sourceChange', this.setSources);
    ArticleStore.on('articleChange', this.setArticles);
  }
  componentWillUnmount() {
    SourcesStore.removeListener('sourceChange', this.setSources);
    ArticleStore.removeListener('articleChange', this.setArticles);
  }

  componentDidMount() {
  }

  // Get all sources from store
  setSources() {
    this.setState({
      sources: SourcesStore.getAll(),
    });
    myActions.getArticles(this.state.sources[0].id);
    // console.log(this.state.sources[0].id);
  }

  // Get All Article from store
  setArticles() {
    this.setState({
      article: ArticleStore.getAll(),
    });
  }

  // Function to trigger Flux Action to Get Sources based on set criteria
  getSources() {
    myActions.getNewSources();
  }

  // Function to trigger Flux Action to Get Articles
  getArticle() {
    myActions.getNewArticles();
  }

  // Search input box
  searchSource(e) {
    const searchString = e.target.value;
    this.setState({
      searchTerm: searchString,
    });
  }

    // Logout from Applicaton
  logout() {
    this.props.logout();
  }

  render() {
    const { article, sources } = this.state;

    const articleComponents = article.map(articleItem => <ArticleComponent key={articleItem.url}{...articleItem} />);

    const sourcesComponents = sources.map((sourcesItem) => {
      const reg = RegExp(this.state.searchTerm, 'gi');
      if (sourcesItem.name.search(reg) !== -1) {
        return <SourcesComponent key={sourcesItem.id}{...sourcesItem} />;
      }
    });

    return (
      <div>
        <Header logout={this.logout.bind(this)} />
        <button hidden onClick={this.logout.bind(this)}>SignOut </button>
        <button hidden onClick={this.getSources.bind(this)}> Get Sources </button>
        <button hidden onClick={this.getArticle.bind(this)}> Get Articles </button>
        <Navbar />
        <div className="row">
          <div className="col-sm-3 col-md-2 sidebar well">
            <h3> All Sources </h3>
            <input type="text" className="form-control" placeholder="Search Sources..." onChange={this.searchSource.bind(this)} />
            <ul className="nav nav-sidebar Source-Container">
              <li className="active"><h4>All Sources <span className="sr-only">(current)</span></h4></li>
              {sourcesComponents}
            </ul>
          </div>
          <div className="container">
            <div className="row article-Container">{articleComponents} </div>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}
