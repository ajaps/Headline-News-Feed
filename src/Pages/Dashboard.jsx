import PropTypes from 'prop-types';
import React from 'react';

import Footer from '../component/Footer.jsx';
import Header from '../component/Header.jsx';
import Navbar from '../component/HeadlineNavbar.jsx';

import ArticleComponent from '../component/ArticleComponent.jsx';
import SourcesComponent from '../component/SourcesComponent.jsx';

import ArticleStore from '../stores/Article';
import SourcesStore from '../stores/Sources';

import * as myActions from '../actions/HeadlineActions';

/**
 * Represents an user's Dsiplay page.
 */
export default class Dashboard extends React.Component {

/**
 * gets sources at the first time of lauching the app
 * binds getArticles and getSources which sets the inital contents gotten from API
 * sests the initial state to mirror the new object in the stores
 * constructor
 * @param {object} props
 */
  constructor(props) {
    super(props);
    this.getSources();

    // calls functions to get default values from store
    this.setSources = this.setSources.bind(this);
    this.logout = this.logout.bind(this);
    this.setArticles = this.setArticles.bind(this);
    this.setLanguage = this.setLanguage.bind(this);
    this.searchSource = this.searchSource.bind(this);

    // Sets state with the available data in the store
    this.state = {
      article: ArticleStore.getAll(),
      sources: SourcesStore.getAll(),
      sortFilter: ArticleStore.getSortAvailable(),
      searchTerm: '',
      url: SourcesStore.getUrl(),
      language: SourcesStore.getSourceUrl.language,
    };
  }

  /**
   * binds Source and Article store to this component
   * @return {void}
   */
  componentWillMount() {
    SourcesStore.on('sourceChange', this.setSources);
    ArticleStore.on('articleChange', this.setArticles);
  }

  /**
   * unbinds Source and Article from this component to prevent memory leaks
   * @return {void}
   */
  componentWillUnmount() {
    SourcesStore.removeListener('sourceChange', this.setSources);
    ArticleStore.removeListener('articleChange', this.setArticles);
  }

  /**
  * changes language preference based on user selection
  * @param {String} lang The preffered language.
  * @returns {void}
  */
  setLanguage(lang) {
    myActions.setLanguage(this.state.url, lang);
  }

  /**
   * sets the state of the component to store values and sends the first source id to generate articles on first load
   * @return {void}
   */
  setSources() {
    this.setState({
      sources: SourcesStore.getAll(),
    });
    myActions.getArticles(ArticleStore.getArticleUrl,
      this.state.sources[0].id, ArticleStore.sortAvailable);
  }

  /**
   * sets the state of the component to store values and sets the filter available for the source
   * @return {void}
   */
  setArticles() {
    this.setState({
      article: ArticleStore.getAll(),
      sortFilter: ArticleStore.getSortAvailable(),
    });
  }

  /**
   * initiate an action to get sources based on the defualt language on first load
   * @return {void}
   */
  getSources() {
    myActions.setLanguage(SourcesStore.getSourceUrl, 'en');
  }

  /**
   * sets the an object in the state to the value of the search string
   * @param {Object} e The calling object property
   * @return {void}
   */
  searchSource(e) {
    const searchString = e.target.value;
    this.setState({
      searchTerm: searchString,
    });
  }

  /**
   * Initiates logout process
   * @return {void}
   */
  logout() {
    this.props.logout();
  }

/**
 * @returns {ReactElement} User dashboard Page.
 */
  render() {
    const { article, sources } = this.state;

    // iterate through article object
    const articleComponents = article.map(articleItem =>
      <ArticleComponent key={articleItem.url}{...articleItem} />);

    // Search through sources, return all if search term is empty
    const sourcesComponents = sources.map((sourcesItem) => {
      const reg = RegExp(this.state.searchTerm, 'gi');
      if (sourcesItem.name.search(reg) !== -1) {
        return <SourcesComponent key={sourcesItem.id}{...sourcesItem} />;
      }
    });

    // Gets all the available languages and sends to header component
    const allLanguages = SourcesStore.getAllLanguages();

    return (
      <div>
        <Header allLanguage={allLanguages} logout={this.logout} url={this.state.url.language}
          setLan={this.setLanguage}/>
        <Navbar sortFilter={this.state.sortFilter} />
        <div className="row">
          <div className="col-sm-3 col-md-2 sidebar well">
            <h3> All Sources </h3>
            <input type="text" className="form-control" placeholder="Search Sources..."
              onChange={this.searchSource} />
            <ul className="nav nav-sidebar Source-Container">
              <li className="active"><h4><span className="sr-only">
                  (current)</span></h4></li>
              {sourcesComponents}
            </ul>
          </div>
          <div className="container well articleCountainerHeight">
            <div className="row article-Container">{articleComponents} </div>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

Dashboard.propTypes = {
  logout: PropTypes.func,
};
