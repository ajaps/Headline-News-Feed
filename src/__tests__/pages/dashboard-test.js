import { shallow } from 'enzyme';
import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import toJson from 'enzyme-to-json';
import Dashboard from '../../pages/Dashboard.jsx';

jest.mock('../../api/NewsApi', () => ({
  getData: () => Promise.resolve('getPromise')
}));
const sortBysAvailable = ['top', 'latest'];

describe('Dashboard page', () => {
  let app;
  it('should render as expected', () => {
    app = shallow(<Dashboard />);
    const tree = toJson(app);
    expect(tree).toMatchSnapshot();
  });

  it('should contain one Article component', () => {
    expect(app.find('ArticleComponent').length).toBe(1);
  });

  it('should contain one Source component', () => {
    expect(app.find('SourcesComponent').length).toBe(1);
  });

  it('should pass dashboard sortfilter state to sortFilter component', () => {
    app.setState({ sortFilter: sortBysAvailable });
    expect(app.exists('<SortFilter sortFilter={Array ["top", "latest",]}/>'))
    .toBe(true);
  });

  it('component should unmount without error', () => {
    const renderedDoc = ReactTestUtils.renderIntoDocument(
      <Dashboard />
    );
    renderedDoc.componentWillUnmount();
  });
});
