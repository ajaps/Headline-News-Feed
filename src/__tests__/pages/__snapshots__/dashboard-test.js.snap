// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard page should render as expected 1`] = `
<div>
  <CategoryComponent
    allCategory={Array []}
    currentcategory={
      Array [
        "all",
      ]
    }
  />
  <SortFilter
    sortFilter={Array []}
  />
  <div
    className="row articlesSourcesContainer"
  >
    <SourcesComponent
      error={null}
      sourceSelected="all"
      sources={Array []}
    />
    <ArticleComponent
      articles={Array []}
      error={null}
    />
  </div>
</div>
`;
